Mathlib/CategoryTheory/Limits/Shapes/Pullback/Pasting.lean

```diff
@@ -39,193 +41,422 @@ namespace CategoryTheory.Limits
+  obtain ⟨l₂, hl₂, hl₂'⟩ := PullbackCone.IsLimit.lift' H (s.fst ≫ g₁) s.snd
+    (by rw [← s.condition, Category.assoc])
+  obtain ⟨l₁, hl₁, hl₁'⟩ := PullbackCone.IsLimit.lift' H' s.fst l₂ (by rw [← hl₂, hi₂])
+  refine ⟨l₁, hl₁, by simp [reassoc_of% hl₁', hl₂'], ?_⟩
+  -- Uniqueness also follows from the universal property of both the small squares.
   intro m hm₁ hm₂
-  apply PushoutCocone.IsColimit.hom_ext H
-  · apply PushoutCocone.IsColimit.hom_ext H'
-    · erw [← Category.assoc, hm₁, hl₂, hl₁]
-    · erw [← Category.assoc, h₂, Category.assoc, hm₂, ← hl₂', ← Category.assoc, ← Category.assoc, ←
-        h₂]
-      rfl
-  · erw [hm₂, hl₂']
+  apply PullbackCone.IsLimit.hom_ext H' (by simpa [hl₁] using hm₁)
+  apply PullbackCone.IsLimit.hom_ext H
+  · dsimp at hm₁
+    rw [Category.assoc, ← hi₂, ← t₁.condition, reassoc_of% hm₁, hl₁', hi₂, hl₂]
+  · simpa [hl₁', hl₂'] using hm₂
 
-/-- Given
+variable (t₁)
 
+/-- Given
 ```
 X₁ - f₁ -> X₂ - f₂ -> X₃
 |          |          |
 i₁         i₂         i₃
 ↓          ↓          ↓
 Y₁ - g₁ -> Y₂ - g₂ -> Y₃
 ```
-
 Then the left square is a pullback if the right square and the big square are.
 -/
-def leftSquareIsPullback (H : IsLimit (PullbackCone.mk _ _ h₂))
-    (H' :
-      IsLimit
-        (PullbackCone.mk _ _
-          (show i₁ ≫ g₁ ≫ g₂ = (f₁ ≫ f₂) ≫ i₃ by
-            rw [← Category.assoc, h₁, Category.assoc, h₂, Category.assoc]))) :
-    IsLimit (PullbackCone.mk _ _ h₁) := by
-  fapply PullbackCone.isLimitAux'
+def leftSquareIsPullback (H : IsLimit t₂) (H' : IsLimit (t₂.pasteHoriz t₁ hi₂)) : IsLimit t₁ := by
+  apply PullbackCone.isLimitAux'
   intro s
-  have : s.fst ≫ g₁ ≫ g₂ = (s.snd ≫ f₂) ≫ i₃ := by
-    rw [← Category.assoc, s.condition, Category.assoc, Category.assoc, h₂]
-  rcases PullbackCone.IsLimit.lift' H' s.fst (s.snd ≫ f₂) this with ⟨l₁, hl₁, hl₁'⟩
-  use l₁
-  use hl₁
-  constructor
+  -- Obtain the induced morphism from the universal property of the big square
+  obtain ⟨l, hl, hl'⟩ := PullbackCone.IsLimit.lift' H' s.fst (s.snd ≫ f₂)
+    (by rw [Category.assoc, ← t₂.condition, reassoc_of% s.condition, ← hi₂])
+  refine ⟨l, hl, ?_, ?_⟩
+  -- To check that `l` is compatible with the projections, we use the universal property of `t₂`
   · apply PullbackCone.IsLimit.hom_ext H
-    · erw [Category.assoc, ← h₁, ← Category.assoc, hl₁, s.condition]
-      rfl
-    · erw [Category.assoc, hl₁']
-      rfl
+    · simp [← s.condition, ← hl, ← t₁.condition, ← hi₂]
+    · simpa using hl'
+  -- Uniqueness of the lift follows from the universal property of the big square
   · intro m hm₁ hm₂
     apply PullbackCone.IsLimit.hom_ext H'
-    · erw [hm₁, hl₁]
-    · erw [hl₁', ← hm₂]
-      exact (Category.assoc _ _ _).symm
+    · simpa [hm₁] using hl.symm
+    · simpa [← hm₂] using hl'.symm
+
+/-- Given that the right square is a pullback, the pasted square is a pullback iff the left
+square is. -/
+def pasteHorizIsPullbackEquiv (H : IsLimit t₂) : IsLimit (t₂.pasteHoriz t₁ hi₂) ≃ IsLimit t₁ :=
```

Comment:
You could use the `wheree` syntax, which shall allow you to introduce variables on the left side of `:=` below:
```suggestion
def pasteHorizIsPullbackEquiv (H : IsLimit t₂) : IsLimit (t₂.pasteHoriz t₁ hi₂) ≃ IsLimit t₁ where
```